pkg/resources/server.go

```diff
@@ -0,0 +1,283 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resources
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/intel/sriov-network-device-plugin/pkg/types"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1beta1"
+)
+
+type resourceServer struct {
+	resourcePool      types.ResourcePool
+	endPoint          string // Socket file
+	resouceNamePrefix string
+	grpcServer        *grpc.Server
+	termSignal        chan bool
+	updateSignal      chan bool
+	stopWatcher       chan bool
+	checkIntervals    int // health check intervals in seconds
+}
+
+func newResourceServer(prefix, suffix string, rp types.ResourcePool) types.ResourceServer {
+	sockName := fmt.Sprintf("%s.%s", rp.GetResourceName(), suffix)
+	return &resourceServer{
+		resourcePool:      rp,
+		endPoint:          sockName,
+		resouceNamePrefix: prefix,
+		grpcServer:        grpc.NewServer(),
+		termSignal:        make(chan bool, 1),
+		updateSignal:      make(chan bool),
+		stopWatcher:       make(chan bool),
+		checkIntervals:    20, // updates every 20 seconds
+	}
+}
+
+func (rs *resourceServer) register() error {
+	kubeletEndpoint := filepath.Join(types.SockDir, types.KubeEndPoint)
+	conn, err := grpc.Dial(kubeletEndpoint, grpc.WithInsecure(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}))
+	if err != nil {
+		glog.Errorf("%s device plugin unable connect to Kubelet : %v", rs.resourcePool.GetResourceName(), err)
+		return err
+	}
+	defer conn.Close()
+	client := pluginapi.NewRegistrationClient(conn)
+
+	request := &pluginapi.RegisterRequest{
+		Version:      pluginapi.Version,
+		Endpoint:     rs.endPoint,
+		ResourceName: fmt.Sprintf("%s/%s", rs.resouceNamePrefix, rs.resourcePool.GetResourceName()),
+	}
+
+	if _, err = client.Register(context.Background(), request); err != nil {
+		glog.Errorf("%s device plugin unable to register with Kubelet : %v", rs.resourcePool.GetResourceName(), err)
+		return err
+	}
+	glog.Infof("%s device plugin registered with Kubelet", rs.resourcePool.GetResourceName())
+	return nil
+}
+
+func (rs *resourceServer) Allocate(ctx context.Context, rqt *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	glog.Infof("Allocate() called with %+v", rqt)
+	resp := new(pluginapi.AllocateResponse)
+	for _, container := range rqt.ContainerRequests {
+		containerResp := new(pluginapi.ContainerAllocateResponse)
+		containerResp.Devices = rs.resourcePool.GetDeviceSpecs(rs.resourcePool.GetDeviceFiles(), container.DevicesIDs)
+		containerResp.Envs = rs.resourcePool.GetEnvs(rs.resourcePool.GetResourceName(), container.DevicesIDs)
```

Comment:
If the network attachment crd is known to the Pod then the associated CRD is also attainable by looking at the resourceName in it. Maybe it's something for a discussion.

> Perhaps we can keep the hardcoded Env name as it is and make the resourceName+DeviceID a dict value of Env.

Can you give an example how it should look like?

Reply:
> If the network attachment crd is known to the Pod then the associated CRD is also attainable by looking at the resourceName in it. Maybe it's something for a discussion.

The resourceName returned in "containerResp.Envs" only contains the suffix (e.g sriov), in multiple endpoints mode, how does Pod get to know which prefix is used(assuming multiple prefix are advertised)?

> 
> > Perhaps we can keep the hardcoded Env name as it is and make the resourceName+DeviceID a dict value of Env.
> 
> Can you give an example how it should look like?

for example:
Env['SRIOV-VF-PCI-ADDR'] = [{"resourceName": "intel.com/sriov", "deviceIDs": "00:01.0 00:01.1"}, {"resourceName": "example.com/sriov", "deviceIDs": "00:02.0 00:02.1"}]

Reply:
@ahalim-intel you're right, I didn't think of that overwrite problem. Then the question remains, how the application running in the Pod can get the PCI address allocated for the Pod. This is a real question from kubevirt community where they use current env['SRIOV-VF-PCI-ADDR'] to get PCI address and passthough that to VM. 
I'm not sure if it's a good design for Pod application to query API and parse the network CRD in order to retrieve the resourceName information, it sounds too much overhead for the application. can you think of a better way for application to get the resourceName?

Reply:
I am not suggesting an application should communicate to API to get this. This information is there, we just need to be aware of the strategy of how this information is being exported into the container Env and make use of it if needed. It's pretty much the same as knowing your CRD names when defining a Pod spec for an application. So the assumption here is we are kind of aware  of what kind of networks we are going to connect to and what resource(name) that network is using.
So, the problem here comes down to knowing the convention of  underlying resourceName of the CRD we are connecting to.
When we are selecting a CRD we should know some details of that CRD for the application to be aware of.

Reply:
I'm open for better alternatives

Reply:
@ahalim-intel the very least, can we use full resource names (with intel.com/ prefix) for the variable names? This will at least help us to not assume the prefix is intel.com/, or that it is present and need to be dealt with at all. It will also allow other DPs to mimic the Intel DP behavior if they need to, without modifications to pod apps.

Reply:
One thing we can do on kubevirt side is to precalculate interface-to-resourceName mapping when creating a pod and pass it off band (environment variable with a well known name?) into the pod. The pod then will extract the mapping and use it to decide which SR-IOV DP managed enviroment variable to use to extract a PCI ID for each VM interface.

We may drop the intel.com/ prefix to construct the PCI ID variable name ourselves, but it would be slightly easier for us (and more compatible with any other solutions that may pop up for device management) if we keep the full resource name for the variables. I understand that environment variables can't have slashes, so perhaps we can replace them with a different character (underscore?)

Reply:
Alternatively, Multus could pass the network-to-resourceName mapping into the pod (todo: check if CNI API supports it). Then we would have the mapping defined in all pods, not just kubevirt virt-launcher pods.

Reply:
I was looking at CNI API description today, specifically at Result format: https://github.com/containernetworking/cni/blob/master/SPEC.md#result and there seems to be no official back-channel from CNI back to pod to pass environment variables or anything like that, so Multus doesn't seem to be a good place for mapping calculation as I originally suggested. On the other side, SR-IOV DP is not the right place either, since it doesn't deal with CRDs.

But I've heard there are plans to add a new mutating hook for SR-IOV DP that would fill in requests / limits based on linked network CRDs. The same hook could also calculate and pass the mapping, via an environment variable or something else.